# OLIST E-COMMERCE DATA ANALYSIS

In [1]:
from pyspark.sql import SparkSession
from pyspark.sql.types import *
from pyspark.sql import functions as F

In [2]:
import pandas as pd
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

In [3]:
spark = SparkSession.builder.appName("Olist Ecommerce EDA").master("local[2]").getOrCreate()

Setting default log level to "WARN".
To adjust logging level use sc.setLogLevel(newLevel). For SparkR, use setLogLevel(newLevel).
26/02/18 04:58:56 WARN NativeCodeLoader: Unable to load native-hadoop library for your platform... using builtin-java classes where applicable


### Loading All Olist Silver Datasets

In [4]:
customers = spark.read.parquet("/opt/examples/datasets/brazilian_ecommerce/silver/olist_customers.parquet")
geolocation = spark.read.parquet("/opt/examples/datasets/brazilian_ecommerce/silver/olist_geolocation.parquet")
order_items = spark.read.parquet("/opt/examples/datasets/brazilian_ecommerce/silver/olist_order_items.parquet")
order_payments = spark.read.parquet("/opt/examples/datasets/brazilian_ecommerce/silver/olist_order_payments.parquet")
order_reviews = spark.read.parquet("/opt/examples/datasets/brazilian_ecommerce/silver/olist_order_reviews.parquet")
products = spark.read.parquet("/opt/examples/datasets/brazilian_ecommerce/silver/olist_products.parquet")
sellers = spark.read.parquet("/opt/examples/datasets/brazilian_ecommerce/silver/olist_sellers.parquet")
orders = spark.read.parquet("/opt/examples/datasets/brazilian_ecommerce/silver/olist_orders.parquet")

In [5]:
# Customers
customers.printSchema()
print(f"Total records: {customers.count()}")
customers.limit(5).toPandas()

root
 |-- customer_id: string (nullable = true)
 |-- customer_unique_id: string (nullable = true)
 |-- customer_zip_code_prefix: string (nullable = true)
 |-- customer_state: string (nullable = true)
 |-- customer_city: string (nullable = true)



Total records: 99441


customer_id                customer_unique_id  \
0  06b8999e2fba1a1fbc88172c00ba8bc7  861eff4711a542e4b93843c6dd7febb0   
1  18955e83d337fd6b2def6b18a428ac77  290c77bc529b7ac935b93aa66c333dc3   
2  4e7b3e00288586ebd08712fdd0374a03  060e732b5b29e8181a18229c7b0b2b5e   
3  b2b6027bc5c5109e529d4dc6358b12c3  259dac757896d24d7702b9acbbff3f3c   
4  4f2d8ab171c80ec8364f7c12e35b23ad  345ecd01c38d18a9036ed96c73b8d066   

  customer_zip_code_prefix customer_state          customer_city  
0                    14409             sp                 franca  
1                    09790             sp  sao bernardo do campo  
2                    01151             sp              sao paulo  
3                    08775             sp        mogi das cruzes  
4                    13056             sp               campinas

In [6]:
# Geolocation 
geolocation.printSchema()
print(f"Total records: {geolocation.count()}")
geolocation.limit(5).toPandas()

root
 |-- geolocation_zip_code_prefix: string (nullable = true)
 |-- geolocation_lat: double (nullable = true)
 |-- geolocation_lng: double (nullable = true)
 |-- geolocation_state: string (nullable = true)
 |-- geolocation_city: string (nullable = true)

Total records: 738332


geolocation_zip_code_prefix  geolocation_lat  geolocation_lng  \
0                       01032       -23.538418       -46.634778   
1                       01050       -23.548723       -46.641603   
2                       01027       -23.537853       -46.630467   
3                       01122       -23.531953       -46.640026   
4                       01122       -23.531495       -46.638858   

  geolocation_state geolocation_city  
0                sp        sao paulo  
1                sp        sao paulo  
2                sp        sao paulo  
3                sp        sao paulo  
4                sp        sao paulo

In [7]:
# Order Items
order_items.printSchema()
print(f"Total records: {order_items.count()}")
order_items.limit(5).toPandas()

root
 |-- order_id: string (nullable = true)
 |-- order_item_id: integer (nullable = true)
 |-- product_id: string (nullable = true)
 |-- seller_id: string (nullable = true)
 |-- shipping_limit_date: timestamp (nullable = true)
 |-- price: decimal(10,2) (nullable = true)
 |-- freight_value: decimal(10,2) (nullable = true)

Total records: 112650


order_id  order_item_id  \
0  00010242fe8c5a6d1ba2dd792cb16214              1   
1  00018f77f2f0320c557190d7a144bdd3              1   
2  000229ec398224ef6ca0657da4fc703e              1   
3  00024acbcdf0a6daa1e931b038114c75              1   
4  00042b26cf59d7ce69dfabb4e55b4fd9              1   

                         product_id                         seller_id  \
0  4244733e06e7ecb4970a6e2683c13e61  48436dade18ac8b2bce089ec2a041202   
1  e5f2d52b802189ee658865ca93d83a8f  dd7ddc04e1b6c2c614352b383efe2d36   
2  c777355d18b72b67abbeef9df44fd0fd  5b51032eddd242adc84c38acab88f23d   
3  7634da152a4610f1595efa32f14722fc  9d7a1d34a5052409006425275ba1c2b4   
4  ac6c3623068f30de03045865e4e10089  df560393f3a51e74553ab94004ba5c87   

  shipping_limit_date   price freight_value  
0 2017-09-19 09:45:35   58.90         13.29  
1 2017-05-03 11:05:13  239.90         19.93  
2 2018-01-18 14:48:30  199.00         17.87  
3 2018-08-15 10:10:18   12.99         12.79  
4 2017-02-13 13:57:51  199.90         18.14

In [8]:
# Order Payments
order_payments.printSchema()
print(f"Total records: {order_payments.count()}")
order_payments.limit(5).toPandas()

root
 |-- order_id: string (nullable = true)
 |-- payment_sequential: integer (nullable = true)
 |-- payment_type: string (nullable = true)
 |-- payment_installments: integer (nullable = true)
 |-- payment_value: decimal(10,2) (nullable = true)

Total records: 103886


order_id  payment_sequential payment_type  \
0  b81ef226f3fe1789b1e8b2acac839d17                   1  credit_card   
1  a9810da82917af2d9aefd1278f1dcfa0                   1  credit_card   
2  25e8ea4e93396b6fa0d3dd708e76c1bd                   1  credit_card   
3  ba78997921bbcdc1373bb41e913ab953                   1  credit_card   
4  42fdf880ba16b47b59251dd489d4441a                   1  credit_card   

   payment_installments payment_value  
0                     8         99.33  
1                     1         24.39  
2                     1         65.71  
3                     8        107.78  
4                     2        128.45

In [107]:
# Order Reviews
order_reviews.printSchema()
print(f"Total records: {order_reviews.count()}")
order_reviews.limit(5).toPandas()

root
 |-- review_id: string (nullable = true)
 |-- order_id: string (nullable = true)
 |-- review_score: integer (nullable = true)
 |-- review_comment_title: string (nullable = true)
 |-- review_comment_message: string (nullable = true)
 |-- review_creation_date: timestamp (nullable = true)
 |-- review_answer_timestamp: timestamp (nullable = true)

Total records: 99224


review_id                          order_id  \
0  7bc2406110b926393aa56f80a40eba40  73fc7af87114b39712e6da79b0a377eb   
1  80e641a11e56f04c1ad469d5645fdfde  a548910a1c6147796b98fdf73dbeba33   
2  228ce5500dc1d8e020d8d1322874b6f0  f9e4b658b201a9f2ecdecbb34bed034b   
3  e64fb393e7b32834bb789ff8bb30750e  658677c97b385a9be170737859d3511b   
4  f7c4243c7fe1938f181bec41a392bdeb  8e6bfb81e283fa7e4f11123a3fb894f1   

   review_score review_comment_title  \
0             4                 None   
1             5                 None   
2             5                 None   
3             5                 None   
4             5                 None   

                              review_comment_message review_creation_date  \
0                                               None           2018-01-18   
1                                               None           2018-03-10   
2                                               None           2018-02-17   
3              recebi bem antes do prazo estipulado.           2017-04-21   
4  parabéns lojas lannister adorei comprar pela i...           2018-03-01   

  review_answer_timestamp  
0     2018-01-18 21:46:59  
1     2018-03-11 03:05:13  
2     2018-02-18 14:36:24  
3     2017-04-21 22:02:06  
4     2018-03-02 10:26:53

In [10]:
# Products
products.printSchema()
print(f"Total records: {products.count()}")
products.limit(5).toPandas()

root
 |-- product_id: string (nullable = true)
 |-- product_name_lenght: integer (nullable = true)
 |-- product_description_lenght: integer (nullable = true)
 |-- product_photos_qty: integer (nullable = true)
 |-- product_weight_g: integer (nullable = true)
 |-- product_length_cm: integer (nullable = true)
 |-- product_height_cm: integer (nullable = true)
 |-- product_width_cm: integer (nullable = true)
 |-- product_category_name: string (nullable = true)

Total records: 32951


product_id  product_name_lenght  \
0  1e9e8ef04dbcff4541ed26657ea517e5                   40   
1  3aa071139cb16b67ca9e5dea641aaa2f                   44   
2  96bd76ec8810374ed1b65e291975717f                   46   
3  cef67bcfe19066a932b7673e239eb23d                   27   
4  9dc1a7de274444849c219cff195d0b71                   37   

   product_description_lenght  product_photos_qty  product_weight_g  \
0                         287                   1               225   
1                         276                   1              1000   
2                         250                   1               154   
3                         261                   1               371   
4                         402                   4               625   

   product_length_cm  product_height_cm  product_width_cm  \
0                 16                 10                14   
1                 30                 18                20   
2                 18                  9                15   
3                 26                  4                26   
4                 20                 17                13   

  product_category_name  
0             perfumery  
1                   art  
2        sports_leisure  
3                  baby  
4            housewares

In [11]:
# Sellers
sellers.printSchema()
print(f"Total records: {sellers.count()}")
sellers.limit(5).toPandas()

root
 |-- seller_id: string (nullable = true)
 |-- seller_zip_code_prefix: string (nullable = true)
 |-- seller_state: string (nullable = true)
 |-- seller_city: string (nullable = true)

Total records: 3095


seller_id seller_zip_code_prefix seller_state  \
0  3442f8959a84dea7ee197c632cb2df15                  13023           sp   
1  d1b65fc7debc3361ea86b5f14c68d2e2                  13844           sp   
2  ce3ad9de960102d0677a81f5d0bb7b2d                  20031           rj   
3  c0f3eea2e14555b6faeea3dd58c1b1c3                  04195           sp   
4  51a04a8a6bdcb23deccc82b0b80742cf                  12914           sp   

         seller_city  
0           campinas  
1         mogi guacu  
2     rio de janeiro  
3          sao paulo  
4  braganca paulista

In [12]:
# Orders
orders.printSchema()
print(f"Total records: {orders.count()}")
orders.limit(5).toPandas()

root
 |-- order_id: string (nullable = true)
 |-- customer_id: string (nullable = true)
 |-- order_status: string (nullable = true)
 |-- order_purchase_timestamp: timestamp (nullable = true)
 |-- order_approved_at: timestamp (nullable = true)
 |-- order_delivered_carrier_date: timestamp (nullable = true)
 |-- order_delivered_customer_date: timestamp (nullable = true)
 |-- order_estimated_delivery_date: timestamp (nullable = true)

Total records: 99441


order_id                       customer_id  \
0  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
1  53cdb2fc8bc7dce0b6741e2150273451  b0830fb4747a6c6d20dea0b8c802d7ef   
2  47770eb9100c2d0c44946d9cf07ec65d  41ce2a54c0b03bf3443c3d931a367089   
3  949d5b44dbf5de918fe9c16f97b45f8a  f88197465ea7920adcdbec7375364d82   
4  ad21c59c0840e6cb83a9ceb5573f8159  8ab97904e6daea8866dbdbc4fb7aad2c   

  order_status order_purchase_timestamp   order_approved_at  \
0    delivered      2017-10-02 10:56:33 2017-10-02 11:07:15   
1    delivered      2018-07-24 20:41:37 2018-07-26 03:24:27   
2    delivered      2018-08-08 08:38:49 2018-08-08 08:55:23   
3    delivered      2017-11-18 19:28:06 2017-11-18 19:45:59   
4    delivered      2018-02-13 21:18:39 2018-02-13 22:20:29   

  order_delivered_carrier_date order_delivered_customer_date  \
0          2017-10-04 19:55:00           2017-10-10 21:25:13   
1          2018-07-26 14:31:00           2018-08-07 15:27:45   
2          2018-08-08 13:50:00           2018-08-17 18:06:29   
3          2017-11-22 13:39:59           2017-12-02 00:28:42   
4          2018-02-14 19:46:34           2018-02-16 18:17:02   

  order_estimated_delivery_date  
0                    2017-10-18  
1                    2018-08-13  
2                    2018-09-04  
3                    2017-12-15  
4                    2018-02-26

### Creating Temporary SQL Views

🎯 Goal: Enable Spark SQL queries for Exploratory Data Analysis (EDA)

In [6]:

# 1. Registering the Customers view
# This view represents unique customer records after the cleaning process.
customers.createOrReplaceTempView("olist_customers")

# 2. Registering the Geolocation view
# Contains latitude/longitude coordinates mapped to zip code prefixes for spatial analysis.
geolocation.createOrReplaceTempView("olist_geolocation")

# 3. Registering the Order Items view
# Details the products, prices, and freight costs associated with each individual order item.
order_items.createOrReplaceTempView("olist_order_items")

# 4. Registering the Order Payments view
# Stores payment method details and installment counts for each transaction.
order_payments.createOrReplaceTempView("olist_order_payments")

# 5. Registering the Order Reviews view
# Contains customer satisfaction scores and review comments.
order_reviews.createOrReplaceTempView("olist_order_reviews")

# 7. Registering the Products view
# Provides metadata for items, including weight, dimensions, and category names.
products.createOrReplaceTempView("olist_products")

# 8. Registering the Sellers view
# Information about the sellers who fulfilled the orders, including their location.
sellers.createOrReplaceTempView("olist_sellers")

# 9. Registering the Orders view
# The core dataset containing order status and timestamps (purchase, approval, delivery).
orders.createOrReplaceTempView("olist_orders")

In [ ]:
# THE TOP 5 CUSTOMER STATES WITH THE MOST ORDERS

In [14]:
spark.sql("""
    SELECT c.customer_state, COUNT(o.order_id) as order_count
    FROM olist_customers c
    JOIN olist_orders o ON c.customer_id = o.customer_id
    GROUP BY c.customer_state
    ORDER BY order_count DESC
    LIMIT 5
""").show()

+--------------+-----------+
|customer_state|order_count|
+--------------+-----------+
|            sp|      41746|
|            rj|      12852|
|            mg|      11635|
|            rs|       5466|
|            pr|       5045|
+--------------+-----------+



In [ ]:
# THE STATES WITH THE LOWEST NUMBER OF ORDERS

In [16]:
spark.sql("""
    SELECT c.customer_state, COUNT(o.order_id) as order_count
    FROM olist_customers c
    JOIN olist_orders o ON c.customer_id = o.customer_id
    GROUP BY c.customer_state
    ORDER BY order_count
    LIMIT 5
""").show()

+--------------+-----------+
|customer_state|order_count|
+--------------+-----------+
|            rr|         46|
|            ap|         68|
|            ac|         81|
|            am|        148|
|            ro|        253|
+--------------+-----------+



In [ ]:
# THE TOP 5 CUSTOMER CITY WITH THE MOST ORDERS

In [21]:
spark.sql("""
    SELECT c.customer_city, COUNT(o.order_id) as order_count
    FROM olist_customers c
    JOIN olist_orders o ON c.customer_id = o.customer_id
    GROUP BY c.customer_city
    ORDER BY order_count DESC
    LIMIT 10
""").show(truncate=False)

+---------------------+-----------+
|customer_city        |order_count|
+---------------------+-----------+
|sao paulo            |15540      |
|rio de janeiro       |6882       |
|belo horizonte       |2773       |
|brasilia             |2131       |
|curitiba             |1521       |
|campinas             |1444       |
|porto alegre         |1379       |
|salvador             |1245       |
|guarulhos            |1189       |
|sao bernardo do campo|938        |
+---------------------+-----------+



In [ ]:
# THE CITIES WITH THE LOWEST NUMBER OF ORDERS

In [22]:
spark.sql("""
    SELECT c.customer_city, COUNT(o.order_id) as order_count
    FROM olist_customers c
    JOIN olist_orders o ON c.customer_id = o.customer_id
    GROUP BY c.customer_city
    ORDER BY order_count
    LIMIT 10
""").show(truncate=False)

+-----------------------+-----------+
|customer_city          |order_count|
+-----------------------+-----------+
|pareci novo            |1          |
|nova guataporanga      |1          |
|paranapoema            |1          |
|redencao da serra      |1          |
|caparao                |1          |
|divino das laranjeiras |1          |
|cerro grande           |1          |
|vermelho               |1          |
|sambaiba               |1          |
|centro novo do maranhao|1          |
+-----------------------+-----------+



In [23]:
spark.sql("""
    SELECT COUNT(DISTINCT customer_city)
    FROM olist_customers 
""").show(truncate=False)

+-----------------------------+
|count(DISTINCT customer_city)|
+-----------------------------+
|4112                         |
+-----------------------------+



In [ ]:
# The total number of cities is 4112.

In [27]:
# Total number of cities with low order volume. (fewer than 500 total orders)
count_result = spark.sql("""
    SELECT COUNT(*) as city_count
    FROM (
        SELECT c.customer_city
        FROM olist_customers c
        JOIN olist_orders o ON c.customer_id = o.customer_id
        GROUP BY c.customer_city
        HAVING COUNT(o.order_id) < 500
    )
""")

count_result.show()

+----------+
|city_count|
+----------+
|      4090|
+----------+



In [28]:
### ORDER REPORTS

In [ ]:
# Total unique number of orders

In [29]:
spark.sql("""
    SELECT COUNT(DISTINCT order_id)
    FROM olist_orders
""").show()

# Total unique number of orders: 99441

+------------------------+
|count(DISTINCT order_id)|
+------------------------+
|                   99441|
+------------------------+



In [ ]:
# The number of delivered orders

In [31]:
spark.sql("""
    SELECT COUNT(order_id)
    FROM olist_orders
    WHERE order_status = 'delivered'
""").show()

# The number of delivered orders: 96478

+---------------+
|count(order_id)|
+---------------+
|          96478|
+---------------+



In [34]:
# THE NUMBER OF UNDELIVERED ORDERS

In [32]:
spark.sql("""
    SELECT order_status, COUNT(DISTINCT order_id) as number_of_orders
    FROM olist_orders
    WHERE order_status != 'delivered'
    GROUP BY order_status
""").show()

+------------+----------------+
|order_status|number_of_orders|
+------------+----------------+
|     shipped|            1107|
|    canceled|             625|
|    approved|               2|
|    invoiced|             314|
|     created|               5|
| unavailable|             609|
|  processing|             301|
+------------+----------------+



In [ ]:
# ORDER VOLUME BY DATE

In [33]:
spark.sql("""
    SELECT DATE(order_purchase_timestamp) as order_date, COUNT(order_id) as num_of_orders_daily
    FROM olist_orders
    GROUP BY order_date
    ORDER BY num_of_orders_daily DESC
""").show()

# November 2017 is the month with the highest order volume, largely driven by Black Friday sales.

+----------+-------------------+
|order_date|num_of_orders_daily|
+----------+-------------------+
|2017-11-24|               1176|
|2017-11-25|                499|
|2017-11-27|                403|
|2017-11-26|                391|
|2017-11-28|                380|
|2018-05-07|                372|
|2018-08-06|                372|
|2018-08-07|                370|
|2018-05-14|                364|
|2018-05-16|                357|
|2018-05-15|                352|
|2018-05-09|                344|
|2017-12-04|                337|
|2018-05-08|                331|
|2017-11-29|                323|
|2018-07-31|                322|
|2018-08-16|                320|
|2018-08-14|                316|
|2018-08-08|                316|
|2018-01-22|                314|
+----------+-------------------+
only showing top 20 rows



In [15]:
# TOTAL REVENUE ONLY DELIVERED ORDERS (ORDER AND ORDER ITEMS)

In [17]:
spark.sql("""
    SELECT SUM(oi.price + oi.freight_value) as total_revenue
    FROM olist_orders o
    JOIN olist_order_items oi ON o.order_id = oi.order_id
    WHERE o.order_status = 'delivered'
""").show(truncate=False)

+-------------+
|total_revenue|
+-------------+
|15419773.75  |
+-------------+



In [ ]:
# TOTAL REVENUE APPROVED AND NOT CANCELED ORDERS (ORDER AND PAYMENTS)

In [20]:
spark.sql("""
    SELECT SUM(p.payment_value) as total_revenue
    FROM olist_orders o
    JOIN olist_order_payments p ON o.order_id = p.order_id
    WHERE o.order_status <> 'canceled' AND order_approved_at IS NOT NULL
""").show(truncate=False)

+-------------+
|total_revenue|
+-------------+
|15862973.82  |
+-------------+



In [22]:
# MONTHLY REVENUE FROM APPROVED ORDERS

In [25]:
spark.sql("""
    SELECT 
        DATE_FORMAT(order_purchase_timestamp, 'yyyy-MM') AS year_month,
        SUM(p.payment_value) as total_revenue
    FROM olist_orders o
    JOIN olist_order_payments p ON o.order_id = p.order_id
    WHERE o.order_status <> 'canceled' AND order_approved_at IS NOT NULL
    GROUP BY year_month
    ORDER BY year_month
""").toPandas()

year_month total_revenue
0     2016-09        136.23
1     2016-10      53915.50
2     2016-12         19.62
3     2017-01     138004.83
4     2017-02     287241.34
5     2017-03     442406.37
6     2017-04     409846.01
7     2017-05     588529.96
8     2017-06     507302.62
9     2017-07     585331.36
10    2017-08     667224.47
11    2017-09     723781.27
12    2017-10     773104.35
13    2017-11    1186940.67
14    2017-12     874695.42
15    2018-01    1109464.00
16    2018-02     984652.59
17    2018-03    1156243.76
18    2018-04    1157336.33
19    2018-05    1149483.77
20    2018-06    1021220.02
21    2018-07    1047422.72
22    2018-08     998504.15
23    2018-09        166.46

In [ ]:
# QAURTERLY REVENUE FROM APPROVED ORDERS

In [26]:
spark.sql("""
    SELECT 
        DATE_FORMAT(order_purchase_timestamp, 'yyyy-Q') AS quarter,
        SUM(p.payment_value) as total_revenue
    FROM olist_orders o
    JOIN olist_order_payments p ON o.order_id = p.order_id
    WHERE o.order_status <> 'canceled' AND order_approved_at IS NOT NULL
    GROUP BY quarter
    ORDER BY quarter
""").toPandas()

quarter total_revenue
0  2016-3        136.23
1  2016-4      53935.12
2  2017-1     867652.54
3  2017-2    1505678.59
4  2017-3    1976337.10
5  2017-4    2834740.44
6  2018-1    3250360.35
7  2018-2    3328040.12
8  2018-3    2046093.33

In [27]:
# TOTAL NUMBER OF ORDERS

In [28]:
spark.sql("""
    SELECT 
        COUNT(DISTINCT order_id) as total_num_orders
    FROM olist_orders 
""").toPandas()

total_num_orders
0             99441

In [30]:
# MONTHLY ORDER VOLUME TRENDS

In [29]:
spark.sql("""
    SELECT 
        DATE_FORMAT(order_purchase_timestamp, 'yyyy-MM') AS year_month,
        COUNT(DISTINCT order_id) as total_num_orders
    FROM olist_orders 
    GROUP BY year_month
    ORDER BY year_month
""").toPandas()

year_month  total_num_orders
0     2016-09                 4
1     2016-10               324
2     2016-12                 1
3     2017-01               800
4     2017-02              1780
5     2017-03              2682
6     2017-04              2404
7     2017-05              3700
8     2017-06              3245
9     2017-07              4026
10    2017-08              4331
11    2017-09              4285
12    2017-10              4631
13    2017-11              7544
14    2017-12              5673
15    2018-01              7269
16    2018-02              6728
17    2018-03              7211
18    2018-04              6939
19    2018-05              6873
20    2018-06              6167
21    2018-07              6292
22    2018-08              6512
23    2018-09                16
24    2018-10                 4

In [ ]:
# SEASONALITY ANALYSIS OF ORDER VOLUME

In [31]:
spark.sql("""
    SELECT 
    month(order_purchase_timestamp) as month_num,
    date_format(order_purchase_timestamp, 'MMMM') as month_name, -- 
    COUNT(order_id) as total_orders
FROM olist_orders
GROUP BY month_num, month_name
ORDER BY month_num
""").toPandas()

month_num month_name  total_orders
0           1    January          8069
1           2   February          8508
2           3      March          9893
3           4      April          9343
4           5        May         10573
5           6       June          9412
6           7       July         10318
7           8     August         10843
8           9  September          4305
9          10    October          4959
10         11   November          7544
11         12   December          5674

In [ ]:
# QUARTERLY ORDER VOLUME TRENDS

In [32]:
spark.sql("""
    SELECT 
        DATE_FORMAT(order_purchase_timestamp, 'yyyy-Q') AS quarter,
        COUNT(DISTINCT order_id) as total_num_orders
    FROM olist_orders
    GROUP BY quarter
    ORDER BY quarter
""").toPandas()

quarter  total_num_orders
0  2016-3                 4
1  2016-4               325
2  2017-1              5262
3  2017-2              9349
4  2017-3             12642
5  2017-4             17848
6  2018-1             21208
7  2018-2             19979
8  2018-3             12820
9  2018-4                 4

In [ ]:
# WEEKLY ORDER VOLUME TRENDS

In [38]:
# Reverts to the legacy datetime parsing policy (pre-Spark 3.0)
spark.conf.set("spark.sql.legacy.timeParserPolicy", "LEGACY")

spark.sql("""
    SELECT 
        date_format(order_purchase_timestamp, 'YYYY-ww') AS iso_year_week,
        COUNT(order_id) AS total_orders
    FROM olist_orders
    GROUP BY iso_year_week
    ORDER BY iso_year_week
""").toPandas()

iso_year_week  total_orders
0        2016-37             2
1        2016-38             2
2        2016-41           258
3        2016-42            65
4        2016-43             1
5        2016-52             1
6        2017-01            40
7        2017-02            72
8        2017-03           180
9        2017-04           350
10       2017-05           427
11       2017-06           559
12       2017-07           435
13       2017-08           373
14       2017-09           469
15       2017-10           592
16       2017-11           623
17       2017-12           646
18       2017-13           564
19       2017-14           578
20       2017-15           467
21       2017-16           532
22       2017-17           691
23       2017-18           745
24       2017-19           793
25       2017-20           904
26       2017-21           834
27       2017-22           832
28       2017-23           834
29       2017-24           778
30       2017-25           646
31       2017-26           727
32       2017-27           855
33       2017-28           939
34       2017-29           972
35       2017-30           915
36       2017-31           992
37       2017-32           952
38       2017-33          1036
39       2017-34           875
40       2017-35          1003
41       2017-36           893
42       2017-37          1174
43       2017-38          1002
44       2017-39           954
45       2017-40          1001
46       2017-41          1075
47       2017-42          1110
48       2017-43          1004
49       2017-44           930
50       2017-45          1187
51       2017-46          1329
52       2017-47          2775
53       2017-48          2255
54       2017-49          1736
55       2017-50          1458
56       2017-51          1052
57       2017-52           862
58       2018-01          1261
59       2018-02          1746
60       2018-03          1786
61       2018-04          1634
62       2018-05          1553
63       2018-06          1613
64       2018-07          1610
65       2018-08          1721
66       2018-09          1904
67       2018-10          1661
68       2018-11          1598
69       2018-12          1741
70       2018-13          1454
71       2018-14          1612
72       2018-15          1516
73       2018-16          1695
74       2018-17          1705
75       2018-18          1730
76       2018-19          1991
77       2018-20          1883
78       2018-21          1015
79       2018-22           991
80       2018-23          1437
81       2018-24          1555
82       2018-25          1410
83       2018-26          1439
84       2018-27          1235
85       2018-28           971
86       2018-29          1615
87       2018-30          1673
88       2018-31          1970
89       2018-32          2067
90       2018-33          1868
91       2018-34          1202
92       2018-35           203
93       2018-36             7
94       2018-37             4
95       2018-38             2
96       2018-39             3
97       2018-40             2
98       2018-42             2

In [39]:
# ORDER VOLUME BY DAY OF WEEK
spark.sql("""
    SELECT 
        date_format(order_purchase_timestamp, 'EEEE') AS day_of_week,
        COUNT(order_id) AS total_orders
    FROM olist_orders
    GROUP BY day_of_week
    ORDER BY total_orders DESC
""").show()

+-----------+------------+
|day_of_week|total_orders|
+-----------+------------+
|     Monday|       16196|
|    Tuesday|       15963|
|  Wednesday|       15552|
|   Thursday|       14761|
|     Friday|       14122|
|     Sunday|       11960|
|   Saturday|       10887|
+-----------+------------+



In [50]:
# TOP 10 PRODUCT CATEGORIES BY ORDER COUNT

In [46]:
spark.sql("""
    SELECT p.product_category_name, COUNT(*) AS num_of_orders
    FROM olist_order_items oi
    JOIN olist_products p ON oi.product_id = p.product_id
    GROUP BY p.product_category_name
    ORDER BY num_of_orders DESC
    LIMIT 10
""").toPandas()

product_category_name  num_of_orders
0         bed_bath_table          11115
1          health_beauty           9670
2         sports_leisure           8641
3        furniture_decor           8334
4  computers_accessories           7827
5             housewares           6964
6          watches_gifts           5991
7              telephony           4545
8           garden_tools           4347
9                   auto           4235

In [ ]:
# TOP 10 PRODUCT CATEGORIES BY TOTAL REVENUE

In [49]:
spark.sql("""
    SELECT p.product_category_name, SUM(oi.price) AS total_revenue
    FROM olist_order_items oi
    JOIN olist_products p ON oi.product_id = p.product_id
    GROUP BY p.product_category_name
    ORDER BY total_revenue DESC
    LIMIT 10
""").toPandas()

product_category_name total_revenue
0          health_beauty    1258681.34
1          watches_gifts    1205005.68
2         bed_bath_table    1036988.68
3         sports_leisure     988048.97
4  computers_accessories     911954.32
5        furniture_decor     729762.49
6             cool_stuff     635290.85
7             housewares     632248.66
8                   auto     592720.11
9           garden_tools     485256.46

In [ ]:
# AVERAGE ORDER VALUE(AOV)

In [51]:
spark.sql("""
    SELECT 
    SUM(price + freight_value) / COUNT(DISTINCT order_id) AS average_order_value
FROM olist_order_items
""").toPandas()

average_order_value
0  160.5776380921492712789

In [ ]:
# AVERAGE ORDER VALUE (AOV) BY PRODUCT CATEGORY

In [54]:
spark.sql("""
    SELECT p.product_category_name, SUM(oi.price + freight_value) / COUNT(DISTINCT oi.order_id) AS average_order_value
    FROM olist_order_items oi
    JOIN olist_products p ON oi.product_id = p.product_id
    GROUP BY p.product_category_name
    ORDER BY average_order_value DESC
""").toPandas()

product_category_name        average_order_value
0                                 computers  1231.84049723756906077348
1     small_appliances_home_oven_and_coffee   632.60946666666666666667
2                         home_appliances_2   484.26384615384615384615
3                agro_industry_and_commerce   398.51906593406593406593
4                       musical_instruments   304.93452229299363057325
5                          small_appliances   302.61679365079365079365
6    kitchen_portables_and_food_preparators   283.46642857142857142857
7                           fixed_telephony   274.57603686635944700461
8                 construction_tools_safety   242.78155688622754491018
9                          air_conditioning   217.48996047430830039526
10                         office_furniture   215.20871956009426551453
11                            watches_gifts   214.26132290184921763869
12                        furniture_bedroom   210.82926315789473684211
13          construction_tools_construction   193.41923796791443850267
14                                 pc_gamer   193.24375000000000000000
15  kitchen_dining_laundry_garden_furniture   186.80794354838709677419
16                               cool_stuff   174.91488160792951541850
17                        home_construction   169.56759183673469387755
18           industry_commerce_and_business   168.80685106382978723404
19                construction_tools_lights   168.36065573770491803279
20                  costruction_tools_tools   163.95824742268041237113
21                    furniture_living_room   163.30938388625592417062
22                                    music   158.79868421052631578947
23                               la_cuisine   158.07615384615384615385
24                   signaling_and_security   153.63735714285714285714
25                                     auto   152.09651270207852193995
26                           consoles_games   148.27233521657250470810
27                             home_confort   147.53662468513853904282
28                                    audio   144.82428571428571428571
29                                     baby   142.72613171577123050260
30                            health_beauty   142.44922363060208239022
31                    security_and_services   141.64500000000000000000
32                             garden_tools   137.93532120523024445708
33                    computers_accessories   136.33642098968455673494
34                      luggage_accessories   135.81235976789168278530
35                 costruction_tools_garden   132.55613402061855670103
36                           sports_leisure   127.98561787564766839378
37                                perfumery   126.22544908285895003163
38                                 pet_shop   125.33064912280701754386
39                                     toys   124.53592382913021101390
40                                  unknown   123.73210199862164024810
41                                      art   119.81504950495049504950
42                           party_supplies   115.00461538461538461538
43        furniture_mattress_and_upholstery   114.94947368421052631579
44                          furniture_decor   113.15901535121724298341
45                           bed_bath_table   110.11879367102049484974
46                               housewares   107.45218558803535010197
47                               cine_photo   106.66861538461538461538
48                          home_appliances   104.93655759162303664921
49                             dvds_blu_ray   101.68457627118644067797
50                             market_place   101.35167857142857142857
51                               stationery    99.93216356555603634790
52                            fashion_shoes    98.17820833333333333333
53                    fashion_male_clothing    96.40910714285714285714
54                   tablets_printing_image    95.29632911392405063291
55                   books_general_interest   

In [55]:
# AVERAGE BASKET SIZE

In [56]:
spark.sql("""
   SELECT 
    COUNT(order_item_id) / COUNT(DISTINCT order_id) AS avg_basket_size
FROM olist_order_items
""").toPandas()

avg_basket_size
0         1.141731

In [ ]:
# AVERAGE REVENUE BY PAYMENT TYPE

In [65]:
spark.sql("""
   SELECT payment_type,
       SUM(payment_value) / COUNT(DISTINCT order_id) AS avg_revenue
   FROM olist_order_payments
   GROUP BY payment_type
   ORDER BY avg_revenue DESC
""").toPandas()

payment_type               avg_revenue
0  credit_card  163.93809803280831318214
1       boleto  145.03443540234532955924
2   debit_card  142.66347513089005235602
3      voucher   98.14714692188308329022
4  not_defined                     0E-20

In [63]:
spark.sql("""
   SELECT *
   FROM olist_order_payments
   WHERE payment_type = 'not_defined'
""").toPandas()

order_id  payment_sequential payment_type  \
0  4637ca194b6387e2d538dc89b124b0ee                   1  not_defined   
1  00b1cb0320190ca0daa2c88b35206009                   1  not_defined   
2  c8c528189310eaa44a745b8d9d26908b                   1  not_defined   

   payment_installments payment_value  
0                     1          0.00  
1                     1          0.00  
2                     1          0.00

In [ ]:
# SELLERS 

In [95]:
spark.sql("""
   SELECT 
        oi.seller_id,
        SUM(oi.price + oi.freight_value) as total_revenue,
        MIN(o.order_purchase_timestamp) as first_order,
        MAX(o.order_purchase_timestamp) as last_order,
        COUNT(DISTINCT o.order_id) as total_orders,
        COUNT(oi.product_id) as total_products
    FROM olist_order_items oi
    JOIN olist_orders o ON oi.order_id = o.order_id
    WHERE o.order_status <> 'canceled'
    GROUP BY oi.seller_id 
    LIMIT 10
""").toPandas()

seller_id total_revenue         first_order  \
0  e63e8bfa530fb16910dd6956e592bb81       1979.17 2017-08-07 09:14:30   
1  0ea22c1cfbdc755f86b9b54b39c16043      15075.31 2016-10-06 18:50:25   
2  8e6cc767478edae941d9bd9eb778d77a       8719.57 2017-01-29 19:08:00   
3  062ce95fa2ad4dfaedfc79260130565f       8721.44 2017-01-26 18:48:30   
4  a49928bcdf77c55c6d6e05e09a9b4ca5      10574.25 2017-06-07 20:32:48   
5  297d5eccd19fa9a83b2630071ff105e4        976.73 2018-07-19 22:48:13   
6  ec8879960bd2221d5c32f8e12f7da711       4881.38 2017-05-16 22:35:38   
7  9c068d10aca38e85c50202e17b4a7e88       3108.83 2017-10-12 09:57:12   
8  ff063b022a9a0aab91bad2c9088760b7      14061.27 2017-02-14 06:59:36   
9  0b64bcdb0784abc139af04077d49a20e        205.81 2018-04-06 19:54:30   

           last_order  total_orders  total_products  
0 2018-08-28 10:44:03            42              43  
1 2018-08-22 18:03:39           237             269  
2 2018-06-13 11:37:39           104             112  
3 2017-09-10 00:14:56            55              60  
4 2018-05-27 19:10:54            98             106  
5 2018-08-23 09:35:05             8               8  
6 2018-04-11 16:50:30            36              36  
7 2018-07-30 14:26:27            17              17  
8 2018-08-20 14:39:22            94              96  
9 2018-07-13 10:56:30             4               4

In [ ]:
# TOTAL NUMBER OF SELLERS WHO HAVE BEEN SELLING FOR AT LEAST THREE MONTHS

In [86]:
spark.sql("""
    WITH seller_stats AS (
        SELECT 
            seller_id,
            MIN(o.order_purchase_timestamp) as first_sale,
            MAX(o.order_purchase_timestamp) as last_sale
        FROM olist_order_items oi
        JOIN olist_orders o on o.order_id = oi.order_id
        WHERE o.order_status <> 'canceled'
        GROUP BY seller_id
    )
    SELECT COUNT(*) as num_of_sellers FROM seller_stats 
    WHERE MONTHS_BETWEEN(last_sale, first_sale) >= 3
""").toPandas()

num_of_sellers
0            1696

In [ ]:
# TOTAL NUMBER OF SELLERS WHO HAVE BEEN SELLING FOR AT LEAST THREE MONTHS BY YEAR

In [91]:
spark.sql("""
    WITH seller_stats AS (
        SELECT 
            seller_id,
            MIN(o.order_purchase_timestamp) as first_order,
            MAX(o.order_purchase_timestamp) as last_order
        FROM olist_order_items oi
        JOIN olist_orders o on o.order_id = oi.order_id
        WHERE o.order_status <> 'canceled'
        GROUP BY seller_id
    )
    SELECT 
        YEAR(first_order) as year,
        COUNT(DISTINCT seller_id) as num_of_sellers  
    FROM seller_stats 
    WHERE MONTHS_BETWEEN(last_order, first_order) >= 3
    GROUP BY year
    ORDER BY year
""").toPandas()

year  num_of_sellers
0  2016             111
1  2017            1176
2  2018             409

In [92]:
# SALES ANALYSIS BASED ON ORDER REVIEWS/SCORES

In [114]:
spark.sql("""
 SELECT 
    or_r.review_score,
    COUNT(DISTINCT o.order_id) as total_num_orders
FROM olist_order_reviews or_r
JOIN olist_orders o ON or_r.order_id = o.order_id
WHERE o.order_status <> 'canceled' AND o.order_approved_at IS NOT NULL
GROUP BY or_r.review_score
ORDER BY or_r.review_score DESC;
""").toPandas()

review_score  total_num_orders
0             5             57000
1             4             19068
2             3              8111
3             2              3103
4             1             10970

In [ ]:
# SALES ANALYSIS BASED ON ORDER REVIEWS/SCORES WITH CATEGORICAL LABELS

In [122]:
spark.sql("""
    SELECT 
        CASE 
            WHEN or_r.review_score = 5 THEN 'Excellent'
            WHEN or_r.review_score = 4 THEN 'Very Good'
            WHEN or_r.review_score = 3 THEN 'Good'
            WHEN or_r.review_score = 2 THEN 'Bad'
            WHEN or_r.review_score = 1 THEN 'Very Bad'
            ELSE 'Unknown' 
        END AS review_label,
        COUNT(DISTINCT o.order_id) as total_num_orders,
        SUM(op.payment_value) as total_revenue,
        ROUND(SUM(op.payment_value) / COUNT(DISTINCT o.order_id), 2) as avg_revenue
    FROM olist_order_reviews or_r
    JOIN olist_orders o ON or_r.order_id = o.order_id
    JOIN olist_order_payments op ON op.order_id = or_r.order_id
    JOIN olist_order_items oi ON oi.order_id = or_r.order_id
    WHERE o.order_status <> 'canceled' AND o.order_approved_at IS NOT NULL
    GROUP BY review_label, or_r.review_score
    ORDER BY or_r.review_score DESC
""").toPandas()

review_label  total_num_orders total_revenue avg_revenue
0    Excellent             56965   10644829.18      186.87
1    Very Good             19048    3640654.56      191.13
2         Good              8076    1640606.12      203.15
3          Bad              3060     766779.55      250.58
4     Very Bad             10509    3345096.71      318.31

In [123]:
spark.sql("""
SELECT 
    review_score,
    COUNT(*) as total_reviews,
    SUM(CASE WHEN review_comment_message IS NULL OR review_comment_message = '' THEN 1 ELSE 0 END) as empty_reviews,
    ROUND(SUM(CASE WHEN review_comment_message IS NULL OR review_comment_message = '' THEN 1 ELSE 0 END) * 100.0 / COUNT(*), 2) as empty_rate_percentage
FROM olist_order_reviews
GROUP BY review_score
ORDER BY review_score DESC;
""").toPandas()

review_score  total_reviews  empty_reviews empty_rate_percentage
0             5          57328          36780                 64.16
1             4          19142          13169                 68.80
2             3           8179           4622                 56.51
3             2           3151           1006                 31.93
4             1          11424           2679                 23.45

**High-Score Contentment:** There is a significant correlation between high satisfaction (5-4 star reviews) and blank comment fields. This suggests that satisfied customers often feel that a maximum star rating is sufficient to communicate their positive experience, resulting in a higher 'silent' majority.

**Low-Score Motivation:** Conversely, customers providing lower ratings (1 or 2 stars) show a much higher engagement in the comment section. These users are typically motivated to document specific issues, such as delivery delays or product defects.

In [ ]:
# CUSTOMER RETENTION AND LOYALTY ANALYSIS

In [124]:
spark.sql("""
    WITH customer_stats AS (
        SELECT 
            c.customer_unique_id,
            COUNT(o.order_id) as num_of_orders
        FROM olist_orders o
        JOIN olist_customers c on o.customer_id = c.customer_id  
        WHERE o.order_status <> 'canceled'
        GROUP BY  c.customer_unique_id
    )
    SELECT 
        COUNT(customer_unique_id) as num_of_loyal_customers
    FROM customer_stats 
    WHERE num_of_orders > 1
""").toPandas()

num_of_loyal_customers
0                    2924

In [127]:
spark.sql("""
    SELECT 
        ROUND(AVG(IF(order_count > 1, 1, 0)) * 100, 2) as loyal_customer_percentage
    FROM (
        SELECT 
            c.customer_unique_id, 
            COUNT(o.order_id) as order_count
        FROM olist_orders o
        JOIN olist_customers c ON o.customer_id = c.customer_id
        WHERE o.order_status <> 'canceled'
        GROUP BY c.customer_unique_id
    )
""").toPandas()

loyal_customer_percentage
0                       3.06

🔍 Our analysis reveals a customer retention rate of **3.06%**, indicating that the vast majority of the Olist customer base consists of one-time buyers.

In [130]:
spark.sql("""
    WITH customer_spending AS (
        -- Calculate total order count and total spending for each unique customer
        SELECT 
            c.customer_unique_id,
            COUNT(DISTINCT o.order_id) as order_count,
            SUM(oi.price) as total_spent
        FROM olist_orders o
        JOIN olist_order_items oi ON o.order_id = oi.order_id
        JOIN olist_customers c ON o.customer_id = c.customer_id
        WHERE o.order_status <> 'canceled'
        GROUP BY c.customer_unique_id
    ),
    revenue_metrics AS (
        -- Aggregate grand total revenue and revenue generated by loyal customers only
        SELECT 
            SUM(total_spent) as grand_total_revenue,
            SUM(IF(order_count > 1, total_spent, 0)) as loyal_customer_revenue
        FROM customer_spending
    )
    -- Final selection to calculate the revenue share percentage of loyal customers
    SELECT 
        ROUND(grand_total_revenue, 2) as total_revenue,
        ROUND(loyal_customer_revenue, 2) as loyal_revenue,
        ROUND((loyal_customer_revenue / grand_total_revenue) * 100, 2) as loyal_revenue_percentage
    FROM revenue_metrics
""").toPandas()

total_revenue loyal_revenue loyal_revenue_percentage
0   13496408.43     750468.43                     5.56

🔍 Loyal customers account for only **5.56%** of total revenue.

In [ ]:
# CANCELLATION RATE

In [132]:
spark.sql("""
    SELECT 
        ROUND(AVG(IF(order_status = 'canceled', 1, 0)) * 100, 2) as cancellation_rate_percentage
    FROM olist_orders
""").toPandas()

cancellation_rate_percentage
0                          0.63

In [ ]:
# BASED ON SELLER CANCELLATION ANALYSIS

In [134]:
spark.sql("""
    SELECT 
        oi.seller_id,
        -- Total number of unique orders per seller
        COUNT(DISTINCT o.order_id) as total_orders,
        -- Count of orders with 'canceled' status
        SUM(IF(o.order_status = 'canceled', 1, 0)) as canceled_orders,
        -- Percentage of canceled orders relative to total orders
        ROUND(AVG(IF(o.order_status = 'canceled', 1, 0)) * 100, 2) as cancellation_rate_percentage
    FROM olist_order_items oi
    JOIN olist_orders o ON oi.order_id = o.order_id
    GROUP BY oi.seller_id
    -- Sorting by total orders to see the most active sellers first
    ORDER BY total_orders DESC
    LIMIT 10
""").toPandas()

seller_id  total_orders  canceled_orders  \
0  6560211a19b47992c3666cc44a7e94c0          1854                8   
1  4a3ca9315b744ce9f8e9374361493884          1806                2   
2  cc419e0650a3c5ba77189a1882b7556a          1706                9   
3  1f50f920176fa81dab994f9023523100          1404                1   
4  da8622b14eb17ae2831f4ac5b9dab84a          1314                0   
5  955fee9216a65b617aa5c0531780ce60          1287                1   
6  7a67c85e85bb2ce8582c35f2203ad736          1160                1   
7  ea8482cd71df3c1969d7b9473ff13abc          1146                1   
8  4869f7a5dfa277a7dca6462dcf3b52b2          1132                1   
9  3d871de0142ce09b7081e2b9d1733cb1          1080                4   

   cancellation_rate_percentage  
0                          0.39  
1                          0.10  
2                          0.51  
3                          0.05  
4                          0.00  
5                          0.07  
6                          0.09  
7                          0.08  
8                          0.09  
9                          0.35

In [ ]:
# TOP SELLING ITEMS

In [136]:
spark.sql("""
    SELECT 
        product_id,
        COUNT(*) as total_items_sold
    FROM olist_order_items 
    GROUP BY product_id
    ORDER BY total_items_sold DESC
    LIMIT 10
""").toPandas()

product_id  total_items_sold
0  aca2eb7d00ea1a7b8ebd4e68314663af               527
1  99a4788cb24856965c36a24e339b6058               488
2  422879e10f46682990de24d770e7f83d               484
3  389d119b48cf3043d311335e499d9c6b               392
4  368c6c730842d78016ad823897a372db               388
5  53759a2ecddad2bb87a079a1f1519f73               373
6  d1c427060a0f73f6b889a5c7c61f2ac4               343
7  53b36df67ebb7c41585e8d54d6772e08               323
8  154e7e31ebfa092203795c972e5804a6               281
9  3dd2a17168ec895c781a9191c1e95ad7               274

In [137]:
# BEST-SELLING PRODUCT CATEGORIES AND SALES TRENDS BY YEAR

In [11]:
spark.sql("""
    SELECT 
        YEAR(o.order_purchase_timestamp) AS sales_year,
        p.product_category_name,
        COUNT(oi.product_id) AS total_items_sold,
        ROUND(SUM(oi.price), 2) AS total_revenue
    FROM olist_order_items oi
    JOIN olist_products p ON oi.product_id = p.product_id
    JOIN olist_orders o ON oi.order_id = o.order_id
    WHERE o.order_status <> 'canceled' AND o.order_approved_at IS NOT NULL
    GROUP BY sales_year, p.product_category_name
    ORDER BY sales_year DESC, total_items_sold DESC
""").toPandas()

sales_year                    product_category_name  total_items_sold  \
0          2018                            health_beauty              5924   
1          2018                           bed_bath_table              5873   
2          2018                    computers_accessories              4683   
3          2018                           sports_leisure              4503   
4          2018                          furniture_decor              4098   
5          2018                               housewares              4020   
6          2018                            watches_gifts              3695   
7          2018                                     auto              2610   
8          2018                                telephony              2330   
9          2018                             garden_tools              1874   
10         2018                              electronics              1853   
11         2018                                     baby              1761   
12         2018                                perfumery              1627   
13         2018                               stationery              1525   
14         2018                                     toys              1472   
15         2018                               cool_stuff              1466   
16         2018                                 pet_shop              1155   
17         2018                         office_furniture               882   
18         2018                 fashion_bags_accessories               879   
19         2018          construction_tools_construction               794   
20         2018                                  unknown               680   
21         2018                          home_appliances               526   
22         2018                      luggage_accessories               491   
23         2018                        home_construction               460   
24         2018                           consoles_games               414   
25         2018                      musical_instruments               396   
26         2018                                     food               381   
27         2018                         small_appliances               353   
28         2018                   books_general_interest               313   
29         2018                                   drinks               287   
30         2018                construction_tools_lights               284   
31         2018                    furniture_living_room               241   
32         2018           industry_commerce_and_business               229   
33         2018                          books_technical               222   
34         2018                                    audio               195   
35         2018                                      art               167   
36         2018  kitchen_dining_laundry_garden_furniture               159   
37         2018                 costruction_tools_garden               157   
38         2018                             home_confort               156   
39         2018                         air_conditioning               156   
40         2018                   signaling_and_security               153   
41         2018               agro_industry_and_commerce               151   
42         2018                        home_appliances_2               148   
43         2018                               food_drink               146   
44         2018                construction_tools_safety               146   
45         2018                             market_place               107   
46         2018                          fixed_telephony               100   
47         2018                       christmas_supplies                90   
48         2018                  costruction_tools_tools                81   
49         2018                                computers                74   
50         2018    small_ap

In [ ]:
# MOST PREFERRED PAYMENT METHOD

In [14]:
spark.sql("""
    SELECT 
        payment_type as payment_method, 
        COUNT(*) as transaction_count
    FROM olist_order_payments
    WHERE payment_type <> 'not_defined'
    GROUP BY 1
    ORDER BY transaction_count DESC
""").toPandas()

payment_method  transaction_count
0    credit_card              76795
1         boleto              19784
2        voucher               5775
3     debit_card               1529

In [ ]:
# MOST PREFERRED PAYMENT METHOD PER PRODUCT CATEGORY

In [9]:
spark.sql("""
    WITH UniqueTransactions AS (
        SELECT DISTINCT 
            p.product_category_name, 
            oi.order_id, 
            op.payment_type
        FROM olist_order_items oi
        JOIN olist_products p ON oi.product_id = p.product_id
        JOIN olist_order_payments op ON oi.order_id = op.order_id
    ),
    PaymentCounts AS (
        SELECT 
            product_category_name, 
            payment_type, 
            COUNT(*) as transaction_count
        FROM UniqueTransactions
        GROUP BY 1, 2
    ),
    RankedPayments AS (
        SELECT 
            *,
            ROW_NUMBER() OVER (PARTITION BY product_category_name ORDER BY transaction_count DESC) as rank
        FROM PaymentCounts
    )
    SELECT 
        product_category_name as category, 
        payment_type as top_payment_method, 
        transaction_count
    FROM RankedPayments
    WHERE rank = 1
    ORDER BY transaction_count DESC
""").toPandas()

category top_payment_method  \
0                            bed_bath_table        credit_card   
1                             health_beauty        credit_card   
2                            sports_leisure        credit_card   
3                           furniture_decor        credit_card   
4                     computers_accessories        credit_card   
5                                housewares        credit_card   
6                             watches_gifts        credit_card   
7                                 telephony        credit_card   
8                                      toys        credit_card   
9                                      auto        credit_card   
10                               cool_stuff        credit_card   
11                             garden_tools        credit_card   
12                                perfumery        credit_card   
13                                     baby        credit_card   
14                              electronics        credit_card   
15                               stationery        credit_card   
16                 fashion_bags_accessories        credit_card   
17                                 pet_shop        credit_card   
18                                  unknown        credit_card   
19                         office_furniture        credit_card   
20                      luggage_accessories        credit_card   
21                           consoles_games        credit_card   
22                          home_appliances        credit_card   
23          construction_tools_construction        credit_card   
24                         small_appliances        credit_card   
25                      musical_instruments        credit_card   
26                        home_construction        credit_card   
27                   books_general_interest        credit_card   
28                    furniture_living_room        credit_card   
29                                     food        credit_card   
30                             home_confort        credit_card   
31                                    audio        credit_card   
32                                   drinks        credit_card   
33                             market_place        credit_card   
34                            fashion_shoes        credit_card   
35                         air_conditioning        credit_card   
36                          books_technical        credit_card   
37  kitchen_dining_laundry_garden_furniture        credit_card   
38                        home_appliances_2        credit_card   
39           industry_commerce_and_business        credit_card   
40                construction_tools_lights        credit_card   
41                               food_drink        credit_card   
42                          fixed_telephony        credit_card   
43                 costruction_tools_garden        credit_card   
44                                computers        credit_card   
45                                      art        credit_card   
46               agro_industry_and_commerce        credit_card   
47                construction_tools_safety        credit_card   
48                       christmas_supplies        credit_card   
49                   signaling_and_security        credit_card   
50                  fashion_underwear_beach        credit_card   
51                    fashion_male_clothing        credit_card   
52                  costruction_tools_tools        credit_card   
53                        furniture_bedroom        credit_card   
54    small_appliances_home_oven_and_coffee        credit_card   
55                   tablets_printing_image        credit_card   
56                             dvds_blu_ray        credit_card   
57                               cine_photo        credit_card   
58                           books_imported        credit_card   
59        furniture_mattress_and_upholstery        credit_card   
60       